### ISSUE TITLE
Python 3: ctypes.c_wchar has a fixed size of 2 bytes whereas py3 strings have variable character length

### ISSUE THREAD

**Issue by leonardder** [nvaccess/nvda#8981]
_Originally posted in https://github.com/nvaccess/nvda/pull/8953#issuecomment-441386821_

In python 3 unicode strings, 32 bit unicode characters are treated as one character in a string.

## Steps to reproduce
Python 2:
```
>>> len(u"👍👍👍")
6
```

Python 3:
```
>>> len(u"� � � ")
3
```

## Expected problems
It is likely that code involving offsets textInfos will break in a major way on python 3, especially for cases where getTextRange will get the requested offsets based on storyText, such as in simple edit controls. In the example above, storyText will be 3 characters long, whereas storyLength will be at least 6. This will result in broken behavior when reading through text in Notepad.

On Windows, c widechars (c_wchar) are 2 bytes in size. On python 2, the size of unicode characters in unicode strings is also two bytes. However, in Python 3, the unicode character length is variable.
[this](https://www.b-list.org/weblog/2017/sep/05/how-python-does-unicode/) is a lovely article on how Python does unicode:

> In Python 3.3 and later, the internal storage of Unicode is now dynamic and chosen on a per-string basis. Here’s how it works:
> 1. Python parses source code on the assumption that it’s UTF-8.
> 2. When it needs to create string objects, Python determines the highest code point in the string, and looks at the size of the encoding needed to store that code point as-is.
> 3. Python then chooses that encoding — which will be one of latin-1, UCS-2, or UCS-4 — to store the string.

### Discussion
@Jsteh said in https://github.com/nvaccess/nvda/pull/8953#issuecomment-441406191: I think we're going to need to have a way to fetch text as UTF-16 bytes arrays, do the work with those and then convert to strings only when returning text for presentation.

This makes sense to do, however when we convert a wchar array to a python bytes array, we'll have to do the string termination by ourselves.

**Comment by leonardder** [comment]
cc @josephsl

**Comment by leonardder** [comment]
@jcsteh: Actually, a problematic point here is that comtypes will also automatically return python unicode strings, for IAccessibleText for example.

Furthermore, observe the following issue:

```
>>> storyText="\ud83d\ud83d\ud83d"
>>> storyText.encode("utf_16_le")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'utf-16-le' codec can't encode character '\ud83d' in position 0: surrogates not allowed
```

Strings where surrogates are used incorrectly are pretty problematic.

**Comment by jcsteh** [comment]
Yeah, I knew comtypes/ctypes would automatically do the conversion, but I figured we could just convert back if necessary. That's pretty wasteful, but I don't really see a feasible alternative. That said, now I'm curious as to what ctypes/comtypes will do with invalid surrogate usage. Does it use surrogatepass internally?

**Comment by leonardder** [comment]
> That said, now I'm curious as to what ctypes/comtypes will do with invalid surrogate usage. Does it use surrogatepass internally?

I have no idea, it seems like it just allows you to use surrogates in a wrong manner, just like you're allowed to create unicode strings with wrongly used surrogates. It just does not allow you to encode them to whatever encoding, but with ctypes, encoding/decoding is not yet involved.

As far as I see, we need to use UTF_16 internally for the following textInfos:

* EditTextInfo
* InputCompositionTextInfo
* appModules.powerpnt.TextFrameTextInfo
* DisplayModelTextInfo
* Virtual buffers

For the following text infos, there is an implementation for _getTextRange that doesn't require us to have an internal UTF-16 encoded representation of the text, or they can just use the offsets in Python 3 unicode strings.

* IA2TextTextInfo
* IA2TextMozilla.FakeEmbeddingTextInfo
* JABTextInfo
* ScintillaTextInfo
* appModules.devenv.VsTextEditPaneTextInfo
* contentRecog.LwrTextInfo: offsets are calculated internally and these text infos are not very detailed (i.e. word is the lowest resolution they support)
* WinConsoleTextInfo

Please correct me if I'm wrong.

**Comment by leonardder** [comment]
Relevant Python Issue:
[create_unicode_buffer() fails on non-BMP strings on Windows](https://bugs.python.org/issue19865)

**Comment by feerrenrut** [comment]
Closed with 15d8374e46f0615ae5d892fbba2ae34d594f224e
